src/components/LMap.vue

```diff
@@ -210,9 +215,24 @@ export default {
     };
 
     onMounted(async () => {
-      const { map, CRS, Icon, latLngBounds, latLng, DomEvent } = await import(
-        "leaflet/dist/leaflet-src.esm"
-      );
+      WINDOW_OR_GLOBAL.L = WINDOW_OR_GLOBAL.L || (await import("leaflet"));
```

Comment:
The reason why I am saying that we lose `tree-shakability` is that `leaflet` do not exports a `module` option in its  `package.json` https://github.com/Leaflet/Leaflet/blob/master/package.json

This means that webpack will not find nor use the `esm` module that is necessary for tree-shaking.

This is a long standing issue with leaflet.

I would propose this:

 - we move `beforeMapMount` outside of `options` because it has a special meaning for us and we ship it directly to leaflet.
- if `beforeMapMount` is present we load from `leaflet` as you are doing otherwise we load from the `esm` file
- we add docs that clearly state that if you pass that you have to import the whole bundle

But to be honest I would prefer the user to load leaflet and attach it to `window` themselves because this complicates the use of the library by a good amount

@mikeu what do you think?

Reply:
I tend to agree that it should be straightforward enough for a library user to import Leaflet when their own component mounts, if that is required by their application.

@juan267 is there a reason that you can't import both Leaflet and your plugin in your sample code at the start of the issue, something like this?
```js
export default {
  components: { ... },
  data() { ... },
  async beforeMount() {
    window.L = await import("leaflet/dist/leaflet-src.esm");
    await import("@geoman-io/leaflet-geoman-free");
    await import("@geoman-io/leaflet-geoman-free/dist/leaflet-geoman.css");
  },
  methods: {
    async mapReady(map) {
      map.pm.addControls({
        position: "topleft",
        drawCircle: false,
      });
    },
  },
};
```

---

In general, wherever Leaflet gets imported to support plugins, I think that it is probably better to use the `esm` file to make sure that plugins don't run into issues where `instanceof` doesn't correctly match classes between Leaflets instances imported from two different files.

@DonNicoJs one other option if we _did_ decide to include this functionality would be to use a separate boolean prop for specifying e.g. `globalLeaflet` or `leafletToWindow` or whatever, that defaulted to false and the user had to opt into. Then there'd be the option to have a `beforeMapMount` hook that didn't load it, and separately the ability to get access to the full Leaflet package without having to pass a dummy hook if you didn't need any other functionality there.

Reply:
@mikeu 

>  is there a reason that you can't import both Leaflet and your plugin in your sample code at the start of the issue, something like this?

I try but there's two reasons why it wont work: 

1. Geoman does not work with the import from `window.L = await import("leaflet/dist/leaflet-src.esm");` it needs to use the oficial import("leaflet"). I guess that this is the case for a lot of plugins
2. I try using the `onBeforeMount` hook but it wont hook geoman correctly to the leaflet instance being used by the Lmap. The only way I could make it work was to defer geoman import until after leaflet was imported on Lmap. Which is the reason to have the `beforeMapMount` hook.

Reply:
@mikeu @DonNicoJs Apparently the reality is that sadly a lot of plugins won't work if the `esm` import for leaflet is being used. Maybe a solution would be to let the user of this library decide is he is willing to let go tree shaking and ssr in favor of compatibility with plugins. For this as @mikeu mentioned above an option could be used to let the user decide which import to use or two versions of this package could be created one with `esm` and the other without it.

Reply:
@juan267 Agreed that we need to find a way to accommodate everyone, I've allocated some time to take care of this during the next WE, I'll do an investigation and will try a few approaches. Sorry for the slow communication here!

Reply:
Update: life got a bit in the way and I had no time to work on this, in the meantime I've managed to successfully hook geoman aside of having this hook, I'll be creating a plugin repo soon for it.

Reply:
@DonNicoJs sorry for asking so much, but did you had the chance to work on this: 

> I've managed to successfully hook geoman aside of having this hook, I'll be creating a plugin repo soon for it.

Reply:
@juan267 no apologies needed, I have to apologise to be so slow recently! Yes I've worked on it and I plan to put out something final during the coming long weekend! Feel free to ping me on Discord directly if you have any pressing need!

Reply:
@juan267 Headsup, the method that I found and was testing is sadly not stable, geoman works but some of the functionality does not work :| I discovered this just at the end of the work when I was testing everything 😢 

I am now torn if to invest more time in this or simply help you finalise this PR, would you be up to add the `import the full leaflet or the esm one` behind a config prop as we discussed with @mikeu ?

Reply:
@DonNicoJs sorry to hear that 😢. Sure I will push updates to this PR and hopefully we can move on from there 👍

Reply:
@DonNicoJs I push the updates. Let me know what you think!